### PR TITLE
Correctly skip colors for nan points given to scatter

### LINKED ISSUE #3489: matplotlib scatter shifts color codes when NaN is present
I want to plot some data with a given color code. For example:

x = np.array([0, 1, 2, 3])
b = np.array([1, 0, 2, 3])
colors = np.array(['g', 'r', 'b', 'y'])
plt.scatter(x, b**2, color=colors)

Great, all the points appear, each one with its color. Now, some data is missing:

plt.figure()
plt.scatter(x, np.log10(b), color=colors)

Here comes the problem: the x=1 data is missing (log(0)=NaN), but the colors are skipping this point and the x=2 point is red, not blue. The solution can be:

y = np.log10(b)
mask = np.isfinite(y)
plt.scatter(x[mask], y[mask], color=colors[mask])

but I feel it so uncomfortable to do this... Any other way? Is this a kind of bug?
Thanks,
Christophe

### PR DESCRIPTION
**Pull request by dstansby**
Fixes #3489. As far as I can tell the only problem was that `colors` was never being masked.

### CODE REVIEW

**Comment by tacaswell** [comment]
attn @efiring as you have been working in this code a lot recently.

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -3996,7 +3996,7 @@ def scatter(self, x, y, s=None, c=None, marker=None, cmap=None, norm=None,
             colors = None  # use cmap, norm after collection is created
 
         # c will be unchanged unless it is the same length as x:
```
Comment (story645) [inline lib/matplotlib/axes/_axes.py]: c and colors will be unchanged (why are c and colors different args?)

File: lib/matplotlib/tests/test_axes.py [hunk 2]
```diff
@@ -4841,3 +4841,13 @@ def test_color_length_mismatch():
...
+
+
+@cleanup
+def test_scatter_color_masking():
+    x = np.array([1, 2, 3])
+    y = np.array([1, np.nan, 3])
+    colors = np.array(['k', 'w', 'k'])
+    s = plt.scatter(x, y, color=colors)
+    facecolors = s.get_facecolors()
+    assert_array_equal(facecolors[1], np.array([0, 0, 0, 1]))
```
Comment (story645) [inline lib/matplotlib/tests/test_axes.py]: probably unnecessary, but what about a test that verifies it's the right facecolor for the right value? Just to ensure no weird shifting.

File: lib/matplotlib/axes/_axes.py (hunk 1 above)
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: That is a question you don't want to ask :smiling_imp: .

File: lib/matplotlib/tests/test_axes.py (hunk 2 above)
Comment (dstansby) [inline lib/matplotlib/tests/test_axes.py]: Is there a way to get the scatter points out from `s`? All I can see that might do it is s.get_paths(), but that seems to give arrays that are definitely not what I'm looking for...

**Comment by dstansby** [comment]
Actually plotting the figure in the test
```
import numpy as np
import matplotlib.pyplot as plt

x = np.array([1, 2, 3])
y = np.array([1, np.nan, 3])
colors = np.array(['k', 'w', 'k'])
s = plt.scatter(x, y, color=colors)
plt.show()
```


gives the utterly weird output:
figure_2
I have no idea why the markers are now different sizes... This is occurring with most recent master merged with my branch. I'll try and look into it tomorrow.

File: lib/matplotlib/axes/_axes.py (hunk 1 above)
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: More seriously, there it pre-dates me and is almost certainly to support back compatibility.  `c` can be a scalar, a color, a sequence of colors or a sequence of scalars.  `colors` can  be (iirc) a single color or a list of colors.

`c` comes from scatter being a scalar mappable under the hood and `colors` comes from scatter returning a collection under the hood.

File: lib/matplotlib/tests/test_axes.py (hunk 2 above)
Comment (story645) [inline lib/matplotlib/tests/test_axes.py]: Ugh, unfortunately haven't found a direct way though their might be one. :( As much as I'm loath to suggest an image test, that seems to be the trick. Though the way to test the error you're seeing below is to use the s.get_edgecolors(), so maybe that'd be good enough?

**Comment by tacaswell** [comment]
Looks like you missed the `edgecolors`?

**Comment by dstansby** [comment]
Okay, I think that's everything masked that needs to be. I couldn't work out whether to mask offsets or not so have left it alone.

File: lib/matplotlib/tests/test_axes.py [hunk 3]
```diff
@@ -4848,9 +4848,12 @@ def test_scatter_color_masking():
...
     colors = np.array(['k', 'w', 'k'])
+    linewidths = np.array([1, 2, 3])
     s = plt.scatter(x, y, color=colors)
 
     facecolors = s.get_facecolors()
     linecolors = s.get_edgecolors()
+    linewidths = s.get_linewidths()
     assert_array_equal(facecolors[1], np.array([0, 0, 0, 1]))
     assert_array_equal(linecolors[1], np.array([0, 0, 0, 1]))
+    assert linewidths[1] == 3
```
Comment (story645) [inline lib/matplotlib/tests/test_axes.py]: also maybe assert facecolors[0], name edgecolors = s.get_edgecolors(),

**Comment by story645** [comment]
Did you generate a new image to verify that it didn't do anything weird? And test with another marker?

File: lib/matplotlib/tests/test_axes.py (hunk 3 above)
Comment (dstansby) [inline lib/matplotlib/tests/test_axes.py]: I think I'm missing something here - should I just rename `facecolors` to `edgecolors`?

**Comment by dstansby** [comment]
Yep, I have generated a new image locally to check nothing funny is happening any more.

I'm baffled that any of the tests passed to be honest given I was never setting linewidths at all... should be fixed now.

**Comment by tacaswell** [comment]
Thanks!  backported as e09a6f7 

@dstansby Greatly appreciate all of the work you have been doing recently!

**Pull request merged**